### ISSUE TITLE
Multilanguage posting

### ISSUE THREAD

**Issue by vinzv** [friendica/friendica#6043]
One thing I liked at Facebook was their multilanguage posting feature:

![grafik](https://user-images.githubusercontent.com/74432/47675619-be6cc700-dbba-11e8-84e2-6a41de110231.png)

![grafik](https://user-images.githubusercontent.com/74432/47675624-c3ca1180-dbba-11e8-843a-17f45f537dad.png)

As I have Friendica contacts with and without German language knowledge it would be cool to have something like that. Of course I could go with two profiles which would be fine. I just thought it would be worth discussing. Maybe it could be done by adding some separator for backwards and other protocols compatibility:

```
[EN] Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua.
---
[DE] At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur
```

**Comment by annando** [comment]
Interesting detail: ActivityPub does support multilanguage posts.

**Comment by JonnyTischbein** [comment]
Do you know if facebook translate it itself or uses a public translate service ?
Mastalab (a client for mastodon) is using [Yandex](https://translate.yandex.com/) for translating toots. If something like this imaginable  ?

**Comment by MrPetovan** [comment]
It isn't reading translation, but writing translation. You have multiple textareas to publish your post in multiple languages at once.

**Comment by vinzv** [comment]
> Do you know if facebook translate it itself or uses a public translate service ?
> Mastalab (a client for mastodon) is using [Yandex](https://translate.yandex.com/) for translating toots. Is something like this imaginable ?

Facebook relies on Bing translations. Maybe something like that could be done with the (paid) deepl.com API. But that's another thing and a bit off topic I think.

**Comment by annando** [comment]
This task will mostly be some UI thing. I wouldn't want to introduce new BBCode commands for this.

**Comment by vinzv** [comment]
I didn't mean to add new BBCode. My idea just was to add some UI in Friendica:
- two or more fields (switchable by a dropdown)
- every field adds a language shortcode to it's content in the DB
- Friendica shows shortcode tagged languages in posts switchable
- other protocols/federating service's UIs would get all languages in one post but separated by shortcodes

It's a bit hacky, I know. ;)

**Comment by AndyHee** [comment]
> You have multiple textareas to publish your post in multiple languages at once.

Yes. Supposedly, the idea is to display the post _only in one language_, i.e. the language of your system or browser settings.

(Arguably not as interesting as reading a post in multiple languages.)

**Comment by MrPetovan** [comment]
We have a prototype for the UI, what about storing and displaying the different versions?

**Comment by MrPetovan** [comment]
Friendica already uses a language detector on each posts and the language information is added to the post metadata.

**Comment by SpcCw** [comment]
> Friendica already uses a language detector on each posts and the language information is added to the post metadata.

Then partly it is already done but without post filtering it is of limited use. The problem (not just in Friendica I suppose) is that if user knows languages A and B, they often post something in A, something in B and something in both (or not language specific). Then their monolingual friends see a ton of posts in languages they don't understand and even if collapsed they still take up space in the feed.

Basically two main things are needed: "do not show language X" and "set language for this post" (as I suppose language detection won't work on memes, links and such).

**Comment by tobiasd** [comment]
Just as clarification, the addon (langfilter) does not detect the language, that is done by Friendica. The addon just takes that information and collapses the posting. As you pointed out the collapsing is not very efficient, but the detection has problems especially with short postings, so I think collapsing is better then not showing because of the false positives.

**Comment by SpcCw** [comment]
> but the detection has problems especially with short postings

I think it is better to use both detection and allowing the user to override it. If user sets in their profile "Ru" as default language and/or sets "Ru" explicitly in the post settings then it is safe to trust this and set language confidence to 100%.

Same should happen if language tag is present: #lang-xx ([#lang-de](https://diasp.org/tags/lang-de), [#lang-ru](https://diasp.org/tags/lang-ru)) - if there is one, then set confidence to 100%.

If post is not tagged explicitly then detection runs and tags it with certain confidence.

Then users can choose to filter out posts in unwanted languages if confidence is higher than certain threshold. Maybe keep the addon too so if someone wants to have the collapse option or both (say, to collapse low confidence posts) they can have it.

**Comment by annando** [comment]
Just a side note: With ActivityPub it is now possible to transmit several language versions of a single post. So the transmission would be possible. The question is now, how to store it and how to handle it on the input side.

**Comment by wiwie** [comment]
I'd also love to see this feature implemented.

I'm wondering what the status of this issue is. Has this been pursued any further? 

@MrPetovan mentioned, that a UI prototype exists. Is there a PR for this somewhere that one could look at? @annando mentioned, that the "input side" needs to be dealt with. What are the issues here exactly? Adding multiple input textboxes is probably not the issue. Does the database structure need to be adapted?

**Comment by MrPetovan** [comment]
We still haven't figured out the storage.

**Comment by annando** [comment]
I guess this can be done with a side table to the `item-content`.

**Comment by wiwie** [comment]
Is this an issue you guys would like help with? Would it make sense if I liked into it or does it probably have a too steep learning curve?

**Comment by MrPetovan** [comment]
I personally have a faint grasp on all the code parts that we would need to change to enable this feature so I definitely think this would be an endeavor for most other people. We probably need a better design prototype which would be more accessible because it wouldn't require any in-depth knowledge of the messy Friendica internals.

**Comment by annando** [comment]
I guess that the backend part is something that I would want to do. But the frontend is some "Hic sunt dracones" for me.

**Comment by annando** [comment]
An easy (quick and dirty) solution would be some additional database field for the additonal languages. This field would then contain a JSON array with language code and content. The primary language would be stored at the same place like ever. And systems like Diaspora would only receive the primary language. But via ActivityPub we could transmit the other languages as well. Question then is of course which systems can handle that.

**Comment by MrPetovan** [comment]
No quick and dirty, let's do it right, it doesn't need to be rushed.

**Comment by annando** [comment]
The more advanced way would be to have some additional table as side table to the items. This would contain the `uri-id` and the language code as primary key. There we would store only the additional languages. So the primary language would still be part of the `item-content` table. The problem here is that the individual languages could contain individual mentions and individual hash tags. And since the mentions do define the delivery process, this would lead to some confusion.

Just imagine, that someone directs a post in the primary language to some German support forum and in the additional languages to some English and some French forum. We would most likely deliver the content only to the German forum.

Or the other way around: Someone is mentioned in the primary language but not in the secondary - but due to language settings the mentioned person only reads the second language. Then the person would be confused about being mentioned without seeing this.

These are some obstacles that need to be defined. And this is a huge task for the front end design to avoid these problems.

**Comment by wiwie** [comment]
I'm not sure, whether my idea is already what you were proposing - in that case don't throw stones :-)

To me a natural way of storing and handling multilingual posts would be to treat each language as an individual post / item. One could store them as multiple separate items in the database, one for each of the contained languages, and add a field to the item table, that links the items for different languages of one multilingual post together -possibly establishing a primary language and linking secondary language items to the one with the primary language.

And the delivery process could be handled equally, by transmitting the different language items separately.

> The more advanced way would be to have some additional table as side table to the items. This would contain the `uri-id` and the language code as primary key. There we would store only the additional languages. So the primary language would still be part of the `item-content` table. The problem here is that the individual languages could contain individual mentions and individual hash tags. And since the mentions do define the delivery process, this would lead to some confusion.
> 
> Just imagine, that someone directs a post in the primary language to some German support forum and in the additional languages to some English and some French forum. We would most likely deliver the content only to the German forum.
> 

I feel above approach would deal with the problem with the different recipients (forums) in different languages. Then we would have one post that goes to the German forum, one to the English and another to the French one. Or possibly all of them to all, but then they would need to handle that, group them together and display them as one multilingual post.

> Or the other way around: Someone is mentioned in the primary language but not in the secondary - but due to language settings the mentioned person only reads the second language. Then the person would be confused about being mentioned without seeing this.
> 
> These are some obstacles that need to be defined. And this is a huge task for the front end design to avoid these problems.

I would use the following logic:

- the friendica UI should show all languages of a multilingual posts, possibly in tabs. 
- if a user receives a multilingual post, the UI should show one of the recipient's languages in the foreground and the remaining ones in the background (possibly tabs).
- If a multilingual post is not available in any of the recipient's languages it should be filtered out (if a language filter is enabled). Otherwise a random language could be shown in foreground, the others in background.
- If someone is mentioned in one of the languages of a multilingual post the UI should show the language in foreground that contains the mention and other languages in the background.

Then, if someone receives a mention the language of the mention is shown - with the chance to interact with it and ask the author to compose it in a language the mentioned recipient understands.

**Comment by annando** [comment]
When you would had separated items then it would be much easier just to chose your current language and to do separate posts. This would be very easy to perform.

**Comment by wiwie** [comment]
Yes right, and in my opinion, for the database part this could also be sufficient. but it should be integrated into the UI in such a way, that those posts are grouped together and not shown separately. When propagated to other federating systems they should also be sent grouped using the facility ActivityPub provides for multilingual posts.

**Comment by annando** [comment]
Concerning AP: I'm not sure if I can assign language attributes to hashtags and I couldn't do different routing per language,

**Comment by wiwie** [comment]
@annando why would they be necessary? Are they federated separately from the post they are in?

**Comment by annando** [comment]
You can have hashtags in one version of the post that you haven't got in the other.

**Comment by wiwie** [comment]
I don't see the problem. Can you describe which scenario you are talking about?

The case where you have differing  hashtags in the different languages of a multilingual post should be solvable if one uses my outlined logic. Then the different languages of the post would be different items with their then implicitly language specific hashtags. Am I mistaken?

**Comment by annando** [comment]
How much AP messages per language would you create? A single language with all the languages included or one message per language? Depending on this there are several obstacles.

**Comment by wiwie** [comment]
I don't know how AP is facilitating multilingual posts. Do you have to be sent in one ap message? Can they be split up and connected via some kind of key?

In what I described above I was mostly thinking about how it could work between friendica instances.

**Comment by annando** [comment]
We are mostly using AP between Friendica instances :-)

In AP you do not only have a "content" field, but also a "contentMap" field. This is an array with all language variants of that post.

Separate postings would need separate identifiers (ID, GUID, URI) per language variant. This would lead to several other problems.

**Comment by Extarys** [comment]
The UI part doesn't concern me as much as how this should be read/sent/retrieved from the server-side (PHP/database). The display part of this will come later.

What should be sent in the POST request? What will be returned from the server?

--------
Is this still something you wish to implement eventually?

**Comment by MrPetovan** [comment]
I can promise you that if you make the UI, we'll make the backend.

**Comment by Extarys** [comment]
Well, I still need to know how the data should be sent to the server.
The form sends `body` for the... body of the post and `title` for... wait for it... the title of the post.

How does that sound:
`body[en-US]`
`body[fr-CA]`

If the user entered a title for those, it'll send the same thing : `title[en-US]` and `title[fr-CA]`.

Is there a function somewhere in PHP I can use to get an array of language with their code? `French Canadian (fr-CA), German (de), English (en-US)`. Not sure how much granular you'd like this to be. In the user settings there is a dropdown with a list of languages but it's just the ISO code, that is also turning me off :laughing: 

- The server also need to check if the language is supported, so the user don't send a body with `body[urmama]` and pollute the database.
- The user can provide a title only in the languages they selected. They won't be able to make a title in Japanese in the current example where only english and french were selected.

Also, how many languages you'd want a post to have? Is there a maximum? Can I translate/adapt my post in 253 languages if I want? (I guess not, the UI will be a mess but..)

On the same note (maybe a different issue), is there a way to get the detected language of a post in a JSON object and not a per-rendered string?

**Comment by MrPetovan** [comment]
> How does that sound:
> `body[en-US]`
> `body[fr-CA]`
>
> If the user entered a title for those, it'll send the same thing : `title[en-US]` and `title[fr-CA]`.

This is perfect.

> Is there a function somewhere in PHP I can use to get an array of language with their code? `French Canadian (fr-CA), German (de), English (en-US)`.

Not in the standard libray. In the Friendica Directory I'm using a static array of language names in the given language, see https://github.com/friendica/friendica-directory/blob/7112c651665adb9eafb2236f36141460d08ccf9c/src/classes/Utils/L10n.php#L16-L155

Not sure why you mentioned 253 languages? The limit placed on the number of languages is a good idea to prevent spam.

**Comment by Extarys** [comment]
@tobiasd wrote :
> As you pointed out the collapsing is not very efficient, but the detection has problems especially with short postings, so I think collapsing is better then not showing because of the false positives.

Can we just disable language detection when a post contain less than X characters/words?

> Not sure why you mentioned 253 languages

Because it's a stupidly high number, you don't know me yet.

I'll see how I can make the JOT model more user-friendly and add this functionality.

**Comment by annando** [comment]
We limited the detection routine to just the languages that Friendica itself supports. So I guess that the results mostly should be okay. In fact you can check the detection results in the menu of each post.

**Comment by gbraad** [comment]
I am not sure what language is given to a message, but when I post from Friendica and receive it on Mastodon I always see a Translate option. Either auto-detection fails, or the wrong language is specified.

![Screenshot from 2022-12-26 00-35-17](https://user-images.githubusercontent.com/1894/209475690-2c253987-692d-4494-8a25-3f399179770e.png)

**Comment by strk** [comment]
> I am not sure what language is given to a message, but when I post from Friendica and receive it on Mastodon I always see a Translate option. Either auto-detection fails, or the wrong language is specified.

I think the encoded language is the language you define as your interface language.
See also https://github.com/mastodon/mastodon/issues/28054 for a way for user to tell in which langauge a post is written, and https://floss.social/@strk@puntarella.party/111471144830945589 (but is in italian) as a workaround to manually see the encoded language.

See also https://friendica.opencloud.lu/display/fb3670be-a8e83623b9a5b502-0b81132e for a possible reason why the message filter doesn't work (that one is probably worth a separate Friendica ticket, for "en" != "en-us")

**Comment by annando** [comment]
@strk We are using a language detection library. With the new version there is also the addon "cld" that introduces a better language detection routine.

**Comment by strk** [comment]
I don't believe in language detection. In general, I believe we shouldn't delegate interpretations to machines when we have the ability of explain ourselves. We know how we are talking better than any automatic system. Consider also the case in which you're just posting an image with rendered text... Or a short word which is the same in multiple languages. Maybe with different meanings too! How hard is it to let the user specify the language ? It is implemneted in Mastodon web and android clients for example.

**Comment by annando** [comment]
Well, I guess that eventually we will add some language selector. But in fact during my work with the new language detector I tested it against millions of posts in my database. I tested it with two different detection libraries and also with the provided language codes (the ones that Mastodon users can assign). In fact there is a significant percentage of users who write in a different language than the one that is provided in the message. I have got the impression that many don't care about this. I guess that it is even worse when you writing in several different languages.

**Comment by strk** [comment]
Many don't care because it's not always very clear that you can set the language, but when you know you can you should be able to override whatever a software detector thinks. I'd welcome a software detector to warn me if he thinks I'm writing in a language and I say I'm writing in another, as long as I can hush it :)

**Comment by strk** [comment]
See how the Mastodon Android app handles this: 


![Screenshot_20231126-082854](https://github.com/friendica/friendica/assets/289678/f28a7753-8594-4229-9a0d-05ba232387bd)

**Comment by strk** [comment]
And this is what you see when you didn't type enough for the detector to work

![Screenshot_20231126-085738](https://github.com/friendica/friendica/assets/289678/5ab69cb5-fad1-4299-a2b4-211f79d6eebe)

**Comment by strk** [comment]
Documentation about the language field in statuses, according to Mastodon: https://docs.joinmastodon.org/methods/statuses/#form-data-parameters the value should be ISO 639 according ro them. It does not specify how many letters. The referenced standard is here: https://www.iso.org/iso-639-language-code

I've seen Mastodon using 2 letters and Mastodon using 2 pairs of 2 (badly supported by Mastodon official app). Anything in the w3c AcrivityPub ?

**Comment by SpcCw** [comment]
I ended up disabling CLD addon. It works really nice when it comes to filtering of incoming posts but since it also triggers to set language of my own posts and comments occasional misfires are more painful. Every now and then it mixes up Belorussian, Bulgarian, Serbian, Russian and Ukrainian languages marking outgoing posts wrong. 

As these languages are becoming more active on Fedi more people are enabling some sort of language filtering and this sometimes results in weird hiccups such as not seeing my posts/comments at all and of course I have no idea when it happens.

I think some kind of workaround is needed, maybe not detecting user posts in any languages outside of those listed in their profile?
